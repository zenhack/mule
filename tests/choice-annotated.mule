#XFAIL?
# The classic example of something hindley-milner can't type check. In mule
# we can, but the annotation is necessary.
fn (choice : all a. a -> a -> a).
  { x = choice {} {}
  , y = choice (fn x. x) (fn x. x)
  }
