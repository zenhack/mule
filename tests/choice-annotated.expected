
Desugared:

	(fn choice
	 (let (choice ((: (all a (a -> (a -> a)))) choice))
	  (fix/record
	   (fn $0
	    (((.= x) (((.= y) {}) ((choice (fn x x)) (fn x x)))) ((choice {}) {}))))))

inferred type:

	(all a
	 (all b
	  (all c
	   (exist d
	    ((d -> (d -> d)) ->
	     (record (types (...c))
	      (values ((x (record (types (...b)) (values ()))) (y (a -> a))))))))))

Runtime term:

	(Lam
	 (0 ()
	  (App
	   ((Lam
	     (0 ()
	      (App
	       ((Fix Record)
	        (Lam
	         (1 ()
	          (Record
	           ((x (App ((App ((Var 1) (Record ()))) (Record ()))))
	            (y
	             (App
	              ((App ((Var 1) (Lam (0 () (Var 0))))) (Lam (0 () (Var 0))))))))))))))
	    (Var 0)))))

Evaluated:

	(Lam
	 (0 ()
	  (App
	   ((Lam
	     (0 ()
	      (App
	       ((Fix Record)
	        (Lam
	         (1 ()
	          (Record
	           ((x (App ((App ((Var 1) (Record ()))) (Record ()))))
	            (y
	             (App
	              ((App ((Var 1) (Lam (0 () (Var 0))))) (Lam (0 () (Var 0))))))))))))))
	    (Var 0)))))
